src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala

```diff
@@ -101,40 +119,80 @@ object AnnexScalaInstance {
     val key = keyBuilder.result()
 
     Option(instanceCache.get(key)).getOrElse {
-      // Copy all the jars to the worker's directory because in a sandboxed world the
-      // jars can go away after the work request, so we can't rely on them sticking around.
-      // This should only happen once per compiler version, so it shouldn't happen often.
-      workRequestJarToWorkerJar.foreach { case (workRequestJar, workerJar) =>
-        this.synchronized {
-          // Check for existence of the file just in case another request is also writing these jars
-          // Copying a file is not atomic, so we don't want to end up in a funky state where two
-          // copies of the same file happen at the same time and cause something bad to happen.
-          if (!Files.exists(workerJar)) {
-            try {
-              Files.createDirectories(workerJar.getParent())
-              Files.copy(workRequestJar, workerJar)
-            } catch {
-              // We do not care if the file already exists
-              case _: FileAlreadyExistsException => {}
-              case e: Throwable                  => throw new Exception("Error adding file to instance cache", e)
+      this.synchronized {
```

Comment:
I'm surprised grabbing a lock on the object like this isn't enough to prevent these sorts of race conditions. If it isn't, can we use locking instead of adding all of this complexity? I have no doubt that this logic is correct, but getting this right seems extremely tricky, and wouldn't a lock be just as performant, while being much simpler to implement?

Reply:
The race condition, as far as I understand it, happens due to the `AnnexScalaInstance` (and the classloaders in it) being shared across compilation requests and then one of the threads using that shared instance being Thread.interrupted at just the wrong point.

Locking won't help you much there either. If you wanted to keep `Thread.interrupt` you'd need to create a new classloader for each request, which means you lose the benefits of the worker strategy because you're class loading + JITing every request.

Reply:
Oh, I see what you're saying. I thought the problem was that multiple work requests were calling this method at once, and while one request was still copying over the file, another request wanted to use the file, despite it not being fully copied over. But yeah, I can see how if a work request is generating the file and is cancelled when it's in the middle of copying, that could cause problems.